core/services/vrf/listener_v2.go

```diff
@@ -619,3 +653,27 @@ func toRequestSet(reqs []pendingRequest) map[string]struct{} {
 	}
 	return s
 }
+
+// GasProofVerification is an upper limit on the gas used for verifying the VRF proof on-chain.
+// It can be used to estimate the amount of LINK needed to fulfill a request.
+const GasProofVerification uint32 = 200_000
+
+// EstimateFeeJuels estimates the amount of link needed to fulfill a request
+// given the callback gas limit, the gas price, and the wei per unit link.
+func EstimateFeeJuels(callbackGasLimit uint32, maxGasPriceWei, weiPerUnitLink *big.Int) *big.Int {
+	maxGasUsed := big.NewInt(int64(callbackGasLimit + GasProofVerification))
+	costWei := new(big.Float).SetInt(
```

Comment:
Why use floats here, aren't we just dealing with integer vals (gas limit, gas price, wei per link etc)?

Reply:
I think because of the eventual division by `weiPerUnitLink`, I think if we do an integer division it might end up truncating, since to convert `weiPerUnitLink` to be `weiPerJuel` we'd have to divide by 10^18. There might be a better way of doing this with only integers though.

Reply:
Wouldn't truncation happen anyway though when you convert back to an integer?

Reply:
There will be a little bit of truncation converting from float -> integer, but not too much. I just tried the same ops but solely with `big.Int` instead of `big.Float` and a fee of 1780216203019246680 is truncated to 1000000000000000000. Maybe we should be using `big.Rat` instead?

Reply:
This is the naive approach though, I think there's a subtler way of doing it with just ints, perhaps with a different order of operations.

Reply:
Oh yeah I see what you're saying..hmm. Would the order of operations in the contract work?

    // (1e18 juels/link) (wei/gas * gas) / (wei/link) = juels
    uint256 paymentNoFee = (1e18 * weiPerUnitGas * (gasAfterPaymentCalculation + startGas - gasleft())) /
      uint256(weiPerUnitLink);

Reply:
+1 for `big.Rat`

Reply:
Ended up implementing Nico's suggestion, which is from the VRFCoordinatorV2 contract. At least we'll have some consistency here 🙂